### PR TITLE
EPL-8414 feat(package): support JSON Schema draft6

### PR DESCRIPTION
**Pull request by michaelcarter-wf**
## Ultimate problem:
- the repo only supports draft4 of json_schema. 
- Since a new draft is published every 6 months or so, it is essential for any good JSON Schema implementation to support multiple versions of the standard.

## How it was fixed:
- add all draft6 capabilities
- allow the user to choose (and auto-detect when they don't) which version of the standard schemas are parsed and validated against. 
i.e.
```dart
await JsonSchema.createSchema(schema, JsonSchemaVersion.draft4);
```
- make the default version draft6 (breaking change)
- test against both draft4 and draft6 test cases.

## Testing suggestions:
- ddev test

## Potential areas of regression:
- Tests should cover any areas of regression.

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on HipChat: InfoSec Forum._

File: test/unit/vm/json_schema/validation_test.dart [hunk 1]
```diff
@@ -72,57 +73,88 @@ void main([List<String> args]) {
...
-            final String validationDescription = validationTest['description'];
-            test('${description} : ${validationDescription}', () {
-              final instance = validationTest['data'];
-              bool validationResult;
-              final bool expectedResult = validationTest['valid'];
-              final checkResult = expectAsync0(() => expect(validationResult, expectedResult));
-              JsonSchema.createSchema(schemaData).then((schema) {
-                validationResult = schema.validate(instance);
-                checkResult();
+  // Directory testSuiteFolder = new Directory('./test/JSON-Schema-Test-Suite/tests/draft6/invalidSchemas');
```
Comment (jamesromo-wf) [inline test/unit/vm/json_schema/validation_test.dart]: Remove dead code?

File: lib/src/json_schema/json_schema.dart [hunk 2]
```diff
@@ -135,7 +171,11 @@ class JsonSchema {
   /// that have interdependencies.
   void _validateAndSetAllProperties() {
     _validateAndSetIndividualProperties();
-    _validateInterdependentProperties();
+
+    // Interdependent properties were removed in draft6
```
Comment (jamesromo-wf) [inline lib/src/json_schema/json_schema.dart]: #nit: period at the end

File: lib/src/json_schema/json_schema.dart [hunk 3]
```diff
@@ -403,11 +515,41 @@ class JsonSchema {
...
+      }
+    }
+    return null;
+  }
+
   /// Whether the maximum of the [JsonSchema] is exclusive.
-  bool get exclusiveMaximum => _exclusiveMaximum == null || _exclusiveMaximum;
+  bool get hasExclusiveMaximum => _exclusiveMaximum ?? _exclusiveMaximumV6 != null;
+
+  /// The value of the exclusiveMaximum for the [JsonSchema], if any exists.
```
Comment (jamesromo-wf) [inline lib/src/json_schema/json_schema.dart]: Looks like this should be `exclusiveMinimum`

File: lib/src/json_schema/json_schema.dart [hunk 4]
```diff
@@ -403,11 +515,41 @@ class JsonSchema {
...
+  num get exclusiveMinimum {
+    // If we're on draft6, the property contains the value, return it.
+    if (schemaVersion == JsonSchemaVersions.draft6) {
+      return _exclusiveMinimumV6;
+
+      // If we're on draft4, the property is a bool, so return the max instead.
```
Comment (jamesromo-wf) [inline lib/src/json_schema/json_schema.dart]: `min`

File: lib/src/json_schema/json_schema.dart [hunk 5]
```diff
@@ -403,11 +515,41 @@ class JsonSchema {
...
+    } else {
+      if (hasExclusiveMinimum) {
+        return _minimum;
+      }
+    }
+    return null;
+  }
 
   /// Whether the maximum of the [JsonSchema] is exclusive.
```
Comment (jamesromo-wf) [inline lib/src/json_schema/json_schema.dart]: `minimum`

File: lib/src/json_schema/utils.dart [hunk 6]
```diff
@@ -87,7 +89,34 @@ class DefaultValidators {
   uriValidator(String uri) {
     try {
       final result = Uri.parse(uri);
-      if (result.path.startsWith('//')) return false;
+      // If a URI has no scheme, it is invalid.
+      if (result.path.startsWith('//') || result.scheme.isEmpty) return false;
+      // If a URI contains spaces, it is invalid.
+      if (uri.contains(' ')) return false;
+      // If a URI contains backslashes, it is invalid
```
Comment (jamesromo-wf) [inline lib/src/json_schema/utils.dart]: period

File: lib/src/json_schema/utils.dart [hunk 7]
```diff
@@ -87,7 +89,34 @@ class DefaultValidators {
...
+      return false;
+    }
+  }
+
+  uriReferenceValidator(String uriReference) {
+    try {
+      Uri.parse(uriReference);
+      // If a URI contains spaces, it is invalid.
+      if (uriReference.contains(' ')) return false;
+      // If a URI contains backslashes, it is invalid
```
Comment (jamesromo-wf) [inline lib/src/json_schema/utils.dart]: period

File: lib/src/json_schema/validator.dart [hunk 8]
```diff
@@ -314,6 +359,11 @@ class Validator {
     final propMustValidate = schema.additionalProperties != null && !schema.additionalProperties;
 
     instance.forEach((k, v) {
+      // Validate property names against the provided schema, if any;
```
Comment (jamesromo-wf) [inline lib/src/json_schema/validator.dart]: Do we want a `;` at the end of this?

File: lib/src/json_schema/validator.dart [hunk 9]
```diff
@@ -384,13 +438,22 @@ class Validator {
   }
 
   void _validate(JsonSchema schema, dynamic instance) {
+    /// If the [JsonSchema] is a bool, always return this value
```
Comment (jamesromo-wf) [inline lib/src/json_schema/validator.dart]: period

File: test/unit/vm/json_schema/validation_test.dart [hunk 10]
```diff
@@ -72,57 +73,88 @@ void main([List<String> args]) {
...
+    if (schemaVersion == JsonSchemaVersions.draft4) {
+      shortSchemaVersion = 'draft4';
+    } else if (schemaVersion == JsonSchemaVersions.draft6) {
+      shortSchemaVersion = 'draft6';
+    }
+
+    allTests.forEach((testEntry) {
+      if (testEntry is File) {
+        group('Validations ($shortSchemaVersion) ${path.basename(testEntry.path)}', () {
+          // Skip these for now - reason shown
```
Comment (jamesromo-wf) [inline test/unit/vm/json_schema/validation_test.dart]: period

File: test/unit/vm/json_schema/validation_test.dart [hunk 11]
```diff
@@ -72,57 +73,88 @@ void main([List<String> args]) {
...
+  ];
+
+  runAllTestsForDraftX(JsonSchemaVersions.draft4, allDraft4, commonSkippedFiles, commonSkippedTests);
+  runAllTestsForDraftX(JsonSchemaVersions.draft6, allDraft6, commonSkippedFiles, commonSkippedTests);
+
+  group('Schema self validation', () {
+    for (final version in JsonSchemaVersions.allVersions) {
+      test('version: $version', () {
+        // Pull in the official schema, verify description and then ensure
+        // that the schema satisfies the schema for schemas
```
Comment (jamesromo-wf) [inline test/unit/vm/json_schema/validation_test.dart]: period

**Comment by jamesromo-wf** [comment]
+10

**Pull request merged**